File: examples/middleware/strands/certification-native.mjs
/** Exact installed Strands operations, native executors, and callback report assertions. */
import assert from 'node:assert/strict';
import { createHash } from 'node:crypto';
import { readFile } from 'node:fs/promises';
import { Agent, FunctionTool, Message, BeforeModelCallEvent, AfterInvocationEvent } from '@strands-agents/sdk';
import { z } from 'zod';
import { createMiddlewareRuntime } from '../../../packages/sdk/typescript/dist/middleware/index.js';
import { withCavemanStrands, withCavemanStrandsModel } from '../../../packages/middleware/typescript/dist/strands.js';
import { startRuntime } from '../../../packages/middleware/conformance/runtime-fixture.mjs';
import { providerFixture, nativeResults, nativeTools, sources, FACT, handlePattern } from './certification-provider.mjs';
import { beginNativeCertification, emitJourney } from './certification-evidence.mjs';
import { nativeParallelGap, nativeCancellationBaseline, cancellationBeforeDispatch } from './upstream-native.mjs';
const cells = JSON.parse(await readFile(new URL('./certification-cells.json', import.meta.url), 'utf8')).cells;
const hash = value => createHash('sha256').update(value).digest('hex');
const normalized = value => JSON.stringify(value).replaceAll(handlePattern, '<OPAQUE_RECOVERY_HANDLE>').replaceAll(/"trackingId":"[^"]+"/g, '"trackingId":"<NATIVE_ID>"');
const reportView = r => ({ status: r.status, reason: r.reason, adapter: r.adapter, replacement_count: r.replacement_count, reused_count: r.reused_count,
  transform_ids: [...r.transform_ids], logical_call_id_present: r.logical_call_id !== null, attempt_id_present: r.attempt_id !== null });
const originals = agent => agent.messages.flatMap(m => m.content).filter(p => p.type === 'toolResultBlock' && p.toolUseId.startsWith('read-'));
const messagesText = result => result.lastMessage.content.filter(p => p.type === 'textBlock').map(p => p.text).join('');

async function run(t, service, cell, mode, lifecycleAction = 'abort') {
  const parallel = cell.method === 'parallel_tool_batch', streamed = ['agent.stream', 'cancel_and_close'].includes(cell.method);
  const fixture = await providerFixture(cell.provider, { parallel, gateText: streamed });
  const reports = [], requests = [], plans = [], diagnostics = [], hooks = [], executions = [], phase = { typed: false };
  const runtime = createMiddlewareRuntime({ endpoint: mode === 'outage' ? 'http://127.0.0.1:1' : service.endpoint, mode: mode === 'off' ? 'off' : 'compress',
    onReport(report) { assert.ok(Object.isFrozen(report)); assert.ok(Object.isFrozen(report.transform_ids)); assert.equal(report.schema_version, 1); assert.ok(!JSON.stringify(report).includes(FACT)); reports.push(report); },
    onDiagnostic(event) { diagnostics.push(event); },
    fetch: async (url, init) => { if (url.endsWith('/optimize')) requests.push(JSON.parse(init.body)); const response = await fetch(url, init); if (url.endsWith('/optimize')) plans.push(await response.clone().json()); return response; } });
  const selectedScope = { namespace: `f09-native-${crypto.randomUUID()}`, session_id: 'session', branch_id: 'main', cache_epoch: '0' };
  let parallelResolve; const barrier = new Promise(resolve => parallelResolve = resolve);
  const tools = ['read_logs', ...(parallel ? ['read_aux'] : [])].map(name => new FunctionTool({ name, description: `Read original ${name}`,
    inputSchema: { type: 'object', properties: { marker: { type: 'string' } }, required: ['marker'] },
    callback: async args => {
      assert.deepEqual(args, { marker: name === 'read_logs' ? 'native-main' : 'native-aux' }); executions.push(name);
      if (parallel) { if (executions.length === 2) parallelResolve(); let timer; try { await Promise.race([barrier, new Promise((_, reject) => timer = setTimeout(() => reject(new Error('Native parallel tool executor serialized the barrier')), 3000))]); } finally { clearTimeout(timer); } }
      return sources[name];
    } }));
  const create = () => new Agent(withCavemanStrands({ model: fixture.model(), tools, retryStrategy: null, toolExecutor: parallel ? 'concurrent' : 'sequential',
    systemPrompt: 'Keep native caller configuration.' }, { runtime, scope: selectedScope }));
  let agent, final = null, firstDeltaBeforeEOF = false, cancelled = false, cancelError = null, snapshot = null, hookResumes = 0, typedResult = null;
  let cancellationBoundary = Promise.resolve(), closedBeforeRelease = null, returnRequested = false, returnedBeforeRelease = null, returned = false;
  const installHook = current => current.addHook(BeforeModelCallEvent, event => { hooks.push(event.type); assert.equal(event.agent, current); });
  try {
    agent = create(); installHook(agent);
    if (mode === 'compress') await runtime.ready();
    if (cell.method === 'every_model_continuation') agent.addHook(AfterInvocationEvent, event => {
      if (!hookResumes) { hookResumes++; fixture.state.resume = true; event.resume = 'Recover the persisted source again.'; }
    });
    if (streamed) {
      const controller = new AbortController(), iterator = agent.stream('Read original logs and recover the omitted detail.', { cancelSignal: AbortSignal.any([controller.signal, t.signal]) }); let text = '';
      try {
        while (true) {
          const next = await iterator.next(); if (next.done) { final = next.value; break; }
          const event = next.value;
          if (event.type === 'modelStreamUpdateEvent' && event.event.type === 'modelContentBlockDeltaEvent' && event.event.delta.type === 'textDelta') {
            text += event.event.delta.text;
            if (text === 'retained-') { assert.equal(fixture.state.gateReleased, false); firstDeltaBeforeEOF = true;
              if (cell.method === 'cancel_and_close') {
                if (lifecycleAction === 'abort') { controller.abort(new Error('native-fixture-cancel')); cancelled = true; }
                else returnRequested = true;
                cancellationBoundary = new Promise(resolve => setTimeout(() => { closedBeforeRelease = fixture.state.closedBeforeEOF;
                  if (returnRequested) returnedBeforeRelease = returned; fixture.release(); resolve(); }, 100));
                if (returnRequested) { await iterator.return(); returned = true; break; }
              } else fixture.release(); }
          }
        }
      } catch (error) { if (!controller.signal.aborted) throw error; cancelError = error.name; }
      finally { await iterator.return(); await cancellationBoundary; }
      if (cell.method === 'cancel_and_close') { assert.equal(lifecycleAction === 'abort' ? cancelled : returnRequested, true); assert.equal(text, 'retained-'); }
      else { assert.equal(text, FACT); assert.equal(messagesText(final), FACT); }
    } else { final = await agent.invoke('Read original logs and recover the omitted detail.', { cancelSignal: t.signal }); assert.equal(messagesText(final), FACT); }
    const initialOriginals = originals(agent); assert.equal(initialOriginals.length, parallel ? 2 : 1);
    for (const original of initialOriginals) assert.equal(original.content[0].text, sources[original.toolUseId === 'read-1' ? 'read_logs' : 'read_aux']);
    const phaseStart = { calls: fixture.state.calls.length, reports: reports.length, plans: plans.length };
    if (cell.method === 'resumed_session') {
      snapshot = JSON.parse(JSON.stringify(agent.takeSnapshot({ preset: 'session' })));
      const originalState = JSON.stringify(snapshot.data.messages);
      await service.restart(); if (mode === 'compress') await runtime.ready();
      const restored = create(); restored.loadSnapshot(snapshot); installHook(restored);
      assert.deepEqual(restored.messages.map(m => m.toJSON()), snapshot.data.messages);
      assert.ok(restored.messages.every(m => m.constructor.name === 'Message' && typeof m.toJSON === 'function'));
      fixture.state.resume = true; final = await restored.invoke('Recover the persisted source again.', { cancelSignal: t.signal });
      assert.equal(messagesText(final), FACT); assert.equal(JSON.stringify(snapshot.data.messages), originalState); agent = restored;
    }
    if (cell.method === 'model.structured_output') {
      fixture.state.structured = true; phase.typed = true;
      const originalState = agent.messages.map(m => m.toJSON());
      const typed = new Agent({ model: withCavemanStrandsModel(fixture.model(), { runtime, scope: { ...selectedScope, branch_id: 'typed', cache_epoch: '1' } }),
        messages: originalState.map(Message.fromJSON), retryStrategy: null }); installHook(typed);
      typedResult = await typed.invoke('Return the typed answer.', { structuredOutputSchema: z.object({ answer: z.number().int() }), cancelSignal: t.signal });
      assert.deepEqual(typedResult.structuredOutput, { answer: 42 }); assert.equal(fixture.state.structuredSeen, 2);
      assert.equal(fixture.state.actions.filter(a => a.type === 'structured').length, 1, 'Native typed validation follows the forced schema continuation');
      assert.deepEqual(agent.messages.map(m => m.toJSON()), originalState);
      assert.ok(reports.slice(phaseStart.reports).every(r => r.replacement_count === 0));
      assert.ok(plans.slice(phaseStart.plans).every(p => p.replacements.length === 0));
    }
    const wire = fixture.state.calls[1].body, wireSource = nativeResults(wire, cell.provider).find(r => r.id === 'read-1').text;
    const recoveryActions = fixture.state.actions.filter(a => a.type.startsWith('recover'));
    assert.equal(executions.length, parallel ? 2 : 1); assert.equal(hooks.length, fixture.state.calls.length);
    assert.equal(reports.length, fixture.state.calls.length, 'Exactly one callback reports each model dispatch');
    assert.equal(runtime.lastReport, reports.at(-1));
    for (const [index, report] of reports.entries()) {
      const count = nativeResults(fixture.state.calls[index].body, cell.provider).filter(r => r.id.startsWith('read-') && r.text.match(handlePattern)).length;
      assert.equal(report.replacement_count, count, 'Callback counts the view sent by the native provider SDK');
      assert.equal(report.adapter, 'strands');
      if (mode === 'off') assert.equal(report.status, 'disabled');
      if (count) { assert.ok(['applied', 'reused'].includes(report.status)); assert.ok(report.transform_ids.length); }
      else assert.ok(['skipped', 'disabled'].includes(report.status));
    }
    if (mode === 'compress') {
      assert.notEqual(wireSource, sources.read_logs); assert.equal(wireSource.includes(FACT), false); assert.ok(recoveryActions.length);
      assert.equal(recoveryActions[0].handle, wireSource.match(handlePattern)[0]);
      const mainSource = originals(agent).find(r => r.toolUseId === 'read-1'); assert.equal(mainSource.content[0].text, sources.read_logs);
      if (cell.method === 'resumed_session' || cell.method === 'every_model_continuation') assert.ok(recoveryActions.some(a => a.type === 'recover_resume'));
      assert.equal(fixture.state.calls.length, parallel ? 4 : ['resumed_session', 'every_model_continuation', 'model.structured_output'].includes(cell.method) ? 5 : 3);
    } else {
      assert.equal(wireSource, sources.read_logs); assert.equal(recoveryActions.length, 0); assert.ok(reports.every(r => r.replacement_count === 0));
      assert.equal(fixture.state.calls.length, cell.method === 'model.structured_output' ? 4 : ['resumed_session', 'every_model_continuation'].includes(cell.method) ? 3 : 2);
      if (mode === 'off') assert.equal(requests.length, 0);
    }
    if (parallel) assert.deepEqual(executions.toSorted(), ['read_aux', 'read_logs']);
    if (cell.method === 'every_model_continuation') assert.equal(hookResumes, 1);
    if (streamed) assert.equal(firstDeltaBeforeEOF, true);
    assert.deepEqual(fixture.state.errors, []);
    return { mode, native_entry_point: cell.method === 'model.structured_output' ? 'Agent.invoke(structuredOutputSchema) -> Model.stream, forced native validation' : `Agent.${cell.method === 'resumed_session' ? 'takeSnapshot/loadSnapshot/invoke' : cell.method === 'every_model_continuation' ? 'invoke/AfterInvocationEvent.resume' : streamed ? 'stream' : 'invoke'}`,
      provider_calls: fixture.state.calls.length, native_tool_calls: executions, source_sha256: hash(sources.read_logs), provider_source_sha256: hash(normalized(wireSource)),
      transformed_request_sha256: hash(normalized(wire)), omitted_fact_absent: !wireSource.includes(FACT), recovery_requests: recoveryActions.length,
      recovered_source_sha256: recoveryActions.length ? hash(sources.read_logs) : null, source_identity_preserved: true,
      native_history_sha256: hash(normalized(agent.messages.map(m => m.toJSON()))), native_message_classes: [...new Set(agent.messages.map(m => m.constructor.name))],
      callback_reports: reports.map(reportView), native_model_hook_calls: hooks.length, first_delta_before_eof: firstDeltaBeforeEOF,
      cancellation: { requested: cancelled, error_type: cancelError, provider_closed_before_release: closedBeforeRelease,
        generator_return_requested: returnRequested, return_completed_before_release: returnedBeforeRelease }, snapshot_roundtrip: snapshot !== null, runtime_restarts: snapshot ? 1 : 0, native_hook_resumes: hookResumes,
      typed: phase.typed ? { output: typedResult.structuredOutput, native_calls: fixture.state.calls.length - phaseStart.calls, callback_reports: reports.slice(phaseStart.reports).map(reportView),
        forced_schema: fixture.state.actions.find(a => a.type === 'structured').schema, recovery_tools: nativeTools(fixture.state.calls.at(-1).body, cell.provider).filter(t => t.name === 'caveman_retrieve').length, replacements: 0 } : null };
  } catch (error) {
    t.diagnostic(`CAVEMAN_MIDDLEWARE_FAILURE ${JSON.stringify({ family: 'F09', cell_id: cell.id, mode, lifecycle_action: lifecycleAction,
      provider_calls: fixture.state.calls.length, native_tool_calls: executions, callback_reports: reports.map(reportView), diagnostics,
      runtime_plans: plans.map(plan => ({ status: plan.status, reason: plan.reason, replacements: plan.replacements?.length ?? 0 })),
      required_journey_complete: false })}`);
    throw error;
  } finally { fixture.release(); await fixture.close(); await runtime.close(); }
}

export async function certifyStrands(t, provider, method) {
  const cell = cells.find(c => c.language === 'typescript' && c.provider === provider && c.method === method); assert.ok(cell);
  const service = await startRuntime(); t.after(service.stop);
  const response = await fetch(service.endpoint + '/caveman/v1/middleware/capabilities'); assert.equal(response.status, 200); await beginNativeCertification(t, await response.json());
  if (provider === 'openai' && method === 'parallel_tool_batch') return nativeParallelGap(t, service, cell);
  const controls = {}; for (const mode of ['compress', 'off', 'outage']) controls[mode] = await run(t, service, cell, mode);
  const active = controls.compress, observed = value => ({ outcome: 'observed', ...value });
  assert.deepEqual(controls.off.native_tool_calls, active.native_tool_calls); assert.deepEqual(controls.outage.native_tool_calls, active.native_tool_calls);
  if (method === 'cancel_and_close') {
    const baseline = await nativeCancellationBaseline(t, provider);
    assert.deepEqual(controls.off.cancellation, baseline); assert.deepEqual(controls.outage.cancellation, baseline); assert.deepEqual(active.cancellation, baseline);
    active.unwrapped_native_cancellation = baseline;
    active.strict_transport_cancellation_passed = baseline.provider_closed_before_release;
    const earlyBaseline = await nativeCancellationBaseline(t, provider, 'return'), earlyControls = {};
    for (const mode of ['compress', 'off', 'outage']) {
      earlyControls[mode] = await run(t, service, cell, mode, 'return');
      assert.deepEqual(earlyControls[mode].cancellation, earlyBaseline);
    }
    active.early_generator_close = { unwrapped: earlyBaseline, ...earlyControls };
    active.strict_transport_early_close_passed = earlyBaseline.provider_closed_before_release;
    active.cancellation_before_dispatch = await cancellationBeforeDispatch(t, service, provider);
  }
  const free = cell.structured_output ? { outcome: 'recovery_free', reason: 'native forced schema output through the public model delegate', replacements: 0, recovery_requests: 0, typed: active.typed } : null;
  emitJourney(t, cell, {
    native_application: observed({ provider, method, native_entry_point: active.native_entry_point }),
    real_tool_result: observed({ native_executions: active.native_tool_calls, source_sha256: active.source_sha256, original_history_preserved: true }),
    transformed_provider_request: free ?? observed({ request_sha256: active.transformed_request_sha256, provider_source_sha256: active.provider_source_sha256, omitted_fact_absent: active.omitted_fact_absent }),
    omitted_fact_requested: free ?? observed({ native_recovery_calls: active.recovery_requests, exact_wire_handle_requested: true }),
    host_executes_exact_recovery: free ?? observed({ source_sha256: active.source_sha256, exact_recovered_sha256: active.recovered_source_sha256, native_executor: 'Strands tool registry' }),
    native_result_history_events_and_call_count: observed(active), off_baseline: observed(controls.off), optimizer_unavailable: observed(controls.outage),
  });
}

File: examples/middleware/strands/_certification_native.py
"""Exact Strands native entry points with real local runtime and provider protocols."""
import asyncio
import copy
import hashlib
import json
import os
import threading
import uuid
from dataclasses import FrozenInstanceError
from pathlib import Path

from pydantic import BaseModel
from strands import Agent, tool
from strands.hooks import BeforeModelCallEvent, AfterInvocationEvent
from strands.event_loop.streaming import process_stream
from strands.tools.executors import ConcurrentToolExecutor
from caveman_cloud.middleware import Scope
from caveman_middleware.strands import with_caveman_agent, with_caveman_model
from evidence_runtime import EvidenceRuntime
from python_fixture import restart_runtime
from _certification_provider import Provider, SOURCES, FACT, HANDLE, native_results, native_tools

CELLS = json.loads(Path(__file__).with_name("certification-cells.json").read_text())["cells"]
TEST_ID = "examples/middleware/strands/test_native.py::AsyncNativeStrands.test_certifies_exact_f09_native_operations"
ENDPOINT = os.environ["CAVEMAN_MIDDLEWARE_ENDPOINT"]


class Answer(BaseModel):
    answer: int


def digest(value):
    return hashlib.sha256(value.encode()).hexdigest()


def normalized(value):
    # Native tracking UUIDs and observed first-byte clocks vary across independent
    # executions. Keep their relationships and validate their real shapes before
    # normalizing only the evidence copy; original messages remain untouched.
    tracking_ids = {}

    def stable(item):
        if isinstance(item, list):
            return [stable(child) for child in item]
        if not isinstance(item, dict):
            return item
        view = {key: stable(child) for key, child in item.items()}
        if item.get("role") in ("user", "assistant", "system") and isinstance(item.get("content"), list):
            if "tracking_id" in item:
                tracking = item["tracking_id"]
                assert isinstance(tracking, str) and str(uuid.UUID(tracking)) == tracking
                view["tracking_id"] = tracking_ids.setdefault(tracking, f"<NATIVE_TRACKING_ID_{len(tracking_ids)}>")
            metrics = item.get("metadata", {}).get("metrics", {})
            if "timeToFirstByteMs" in metrics:
                assert isinstance(metrics["timeToFirstByteMs"], (int, float)) and metrics["timeToFirstByteMs"] >= 0
                view["metadata"]["metrics"]["timeToFirstByteMs"] = "<OBSERVED_NONNEGATIVE_FIRST_BYTE_MS>"
        return view

    return HANDLE.sub("<OPAQUE_RECOVERY_HANDLE>", json.dumps(stable(value), ensure_ascii=False, separators=(",", ":"), sort_keys=True))


def report_view(report):
    return {"status": report.status, "reason": report.reason, "adapter": report.adapter, "replacement_count": report.replacement_count,
            "reused_count": report.reused_count, "transform_ids": list(report.transform_ids),
            "logical_call_id_present": report.logical_call_id is not None, "attempt_id_present": report.attempt_id is not None}


def original_results(agent):
    return [p["toolResult"] for m in agent.messages for p in m["content"] if "toolResult" in p and p["toolResult"]["toolUseId"].startswith("read-")]


async def consume_agent(test, agent, server, *, action=None):
    cancel_signal = threading.Event()
    stream = agent.stream_async("Read original logs and recover the omitted detail.", cancel_signal=cancel_signal)
    text, event_types, result = "", [], None
    boundary, closed, returned_before, returned, return_requested, error_type = None, None, None, False, False, None

    async def release_boundary():
        nonlocal closed, returned_before
        await asyncio.sleep(0.1)
        closed = server.closed_before_eof
        if return_requested:
            returned_before = returned
        server.release.set()

    try:
        async for event in stream:
            event_types.extend(event)
            if "result" in event:
                result = event["result"]
            if "data" in event:
                text += event["data"]
                if text == "retained-":
                    test.assertFalse(server.release.is_set(), "native consumer saw a delta before provider EOF")
                    if action:
                        if action == "abort":
                            cancel_signal.set()
                        else:
                            return_requested = True
                        boundary = asyncio.create_task(release_boundary())
                        if return_requested:
                            await stream.aclose(); returned = True
                            break
                    else:
                        server.release.set()
    except asyncio.CancelledError:
        if not cancel_signal.is_set():
            raise
        error_type = "CancelledError"
    finally:
        await stream.aclose()
        if boundary:
            await boundary
    test.assertEqual(text, "retained-" if action else FACT)
    if not action:
        test.assertEqual(result.message["content"], [{"text": FACT}])
    return {"result": result, "event_types": sorted(set(event_types)), "first_delta_before_eof": True,
            "lifecycle": {"requested": cancel_signal.is_set(), "error_type": error_type, "provider_closed_before_release": closed,
                          "generator_return_requested": return_requested, "return_completed_before_release": returned_before}}


async def direct_model_stream(test, agent, server):
    """A bounded public Model caller uses the installed Strands parser and tool executor."""
    server.resume = True
    server.release.clear()
    stream_calls, native_recoveries, event_types, first = 0, 0, [], False
    while stream_calls < 2:
        stream_calls += 1
        specs = agent.tool_registry.get_all_tool_specs()
        message, stop = None, None
        async for event in process_stream(agent.model.stream(agent.messages, specs, agent.system_prompt,
                                                            invocation_state={"agent": agent, "native_marker": "preserved"})):
            event_types.extend(event)
            if event.get("data") == "retained-":
                test.assertFalse(server.release.is_set()); first = True; server.release.set()
            if "stop" in event:
                stop, message, _, _ = event["stop"]
        test.assertIsNotNone(message)
        if stop == "end_turn":
            test.assertEqual(message["content"], [{"text": FACT}])
            agent.messages.append(message)
            break
        test.assertEqual(stop, "tool_use")
        call = next(p["toolUse"] for p in message["content"] if "toolUse" in p)
        test.assertEqual(call["name"], "caveman_retrieve")
        test.assertEqual(call["input"]["handle"], server.handles["read-1"])
        executor = agent.tool_registry.registry[call["name"]]
        native_result = None
        async for event in executor.stream(call, {"agent": agent, "native_marker": "preserved"}):
            if isinstance(event, dict) and "tool_result" in event:
                native_result = event["tool_result"]
        test.assertIsNotNone(native_result)
        test.assertEqual(json.loads(native_result["content"][0]["text"])["text"], SOURCES["read_logs"])
        native_recoveries += 1
        agent.messages.extend([message, {"role": "user", "content": [{"toolResult": native_result}]}])
    test.assertTrue(first)
    return {"direct_model_stream_calls": stream_calls, "native_decorated_executor_calls": native_recoveries,
            "native_parser": "strands.event_loop.streaming.process_stream", "first_delta_before_eof": first, "native_event_types": sorted(set(event_types))}


async def run(test, cell, mode, *, lifecycle_action="abort", unwrapped=False):
    method, protocol = cell["method"], cell["provider"]
    parallel, streamed = method == "parallel_tool_batch", method in ("agent.stream_async", "cancel_and_close")
    server = Provider(protocol, parallel=parallel, gate_text=streamed)
    reports, diagnostics, hooks, executions, source_refs, message_refs = [], [], [], [], {}, {}
    runtime = EvidenceRuntime(endpoint="http://127.0.0.1:1" if mode == "outage" else ENDPOINT,
                              mode="off" if mode == "off" else "compress", on_report=reports.append, on_diagnostic=diagnostics.append)
    selected_scope = Scope("f09-native-" + uuid.uuid4().hex, "session")
    barrier = asyncio.Event()

    def make_tool(name):
        @tool(name=name, description="Read original " + name)
        async def read(marker: str) -> str:
            test.assertEqual(marker, "native-main" if name == "read_logs" else "native-aux")
            executions.append(name)
            if parallel:
                if len(executions) == 2:
                    barrier.set()
                await asyncio.wait_for(barrier.wait(), 3)
            return SOURCES[name]
        return read

    tools = [make_tool(name) for name in (["read_logs", "read_aux"] if parallel else ["read_logs"])]

    def create(messages=None):
        options = {"model": server.model(), "tools": tools, "messages": messages or [], "system_prompt": "Keep native caller configuration.",
                   "callback_handler": None, "tool_executor": ConcurrentToolExecutor()}
        agent = Agent(**(options if unwrapped else with_caveman_agent(options, runtime=runtime, scope=selected_scope)))

        def before(event):
            test.assertIs(event.agent, agent)
            hooks.append(type(event).__name__)
            for message in agent.messages:
                tracking = message.get("tracking_id")
                if tracking is not None:
                    test.assertEqual(str(uuid.UUID(tracking)), tracking)
                    key = (id(agent), id(message))
                    if key in message_refs:
                        test.assertEqual(tracking, message_refs[key], "native message tracking ID remains stable")
                    message_refs[key] = tracking
            for original in original_results(agent):
                key = (id(agent), original["toolUseId"])
                if key in source_refs:
                    test.assertIs(original, source_refs[key], "native agent history retains the original result object")
                source_refs[key] = original
        agent.add_hook(before, BeforeModelCallEvent)
        return agent

    agent, hook_resumes, details = None, 0, {}
    try:
        if mode == "compress" and not unwrapped:
            await asyncio.to_thread(runtime.ready)
        agent = create()
        if method == "every_model_continuation":
            def resume(event):
                nonlocal hook_resumes
                if not hook_resumes:
                    hook_resumes += 1; server.resume = True
                    event.resume = "Recover the persisted source again."
            agent.add_hook(resume, AfterInvocationEvent)
        if method == "agent.__call__":
            result = await asyncio.to_thread(agent, "Read original logs and recover the omitted detail.")
            test.assertEqual(result.message["content"], [{"text": FACT}])
        elif streamed:
            details = await consume_agent(test, agent, server, action=lifecycle_action if method == "cancel_and_close" else None)
            result = details.pop("result")
        else:
            result = await agent.invoke_async("Read original logs and recover the omitted detail.")
            test.assertEqual(result.message["content"], [{"text": FACT}])
        for original in original_results(agent):
            test.assertEqual(original["content"][0]["text"], SOURCES["read_logs" if original["toolUseId"] == "read-1" else "read_aux"])
        phase_start = {"calls": len(server.calls), "reports": len(reports), "plans": len(runtime.plans)}
        if method == "model.stream":
            details.update(await direct_model_stream(test, agent, server))
        elif method == "model.structured_output":
            server.structured = True
            before = copy.deepcopy(agent.messages)
            typed_model = with_caveman_model(server.model(), runtime=runtime, scope=Scope(selected_scope.namespace, selected_scope.session_id, "typed", "1"))
            typed_events = [event async for event in typed_model.structured_output(Answer, agent.messages, agent.system_prompt)]
            test.assertEqual(typed_events[-1]["output"], Answer(answer=42))
            test.assertEqual(agent.messages, before)
            test.assertEqual(len(server.calls) - phase_start["calls"], 1)
            test.assertTrue(all(report.replacement_count == 0 for report in reports[phase_start["reports"]:]))
            test.assertTrue(all(not plan.replacements for _, plan in runtime.plans[phase_start["plans"]:]))
            details["typed"] = {"output": typed_events[-1]["output"].model_dump(), "native_calls": 1,
                "native_iterator_events": len(typed_events), "native_event_types": sorted({key for event in typed_events for key in event}),
                "schema": next(action["schema"] for action in server.actions if action["type"] == "structured"), "replacements": 0,
                "callback_reports": [report_view(report) for report in reports[phase_start["reports"]:]]}
        elif method == "resumed_session":
            persisted = json.dumps(agent.messages, ensure_ascii=False)
            await asyncio.to_thread(restart_runtime)
            if mode == "compress":
                await asyncio.to_thread(runtime.ready)
            agent = create(json.loads(persisted)); test.assertEqual(json.dumps(agent.messages, ensure_ascii=False), persisted)
            server.resume = True
            result = await agent.invoke_async("Recover the persisted source again.")
            test.assertEqual(result.message["content"], [{"text": FACT}])
            details["native_history_json_roundtrip"] = True; details["runtime_process_restarts"] = 1
        wire = server.calls[1]["body"]
        wire_source = next(r["text"] for r in native_results(wire, protocol) if r["id"] == "read-1")
        recoveries = [action for action in server.actions if action["type"].startswith("recover")]
        test.assertEqual(len(executions), 2 if parallel else 1)
        if unwrapped:
            test.assertEqual(reports, [])
        else:
            test.assertEqual(len(reports), len(server.calls), "one callback for every native model dispatch")
            test.assertIs(runtime.last_report, reports[-1])
            for index, report in enumerate(reports):
                test.assertEqual(report.schema_version, 1); test.assertIsInstance(report.transform_ids, tuple)
                with test.assertRaises((FrozenInstanceError, AttributeError)):
                    report.status = "forged"
                test.assertNotIn(FACT, repr(report)); test.assertEqual(report.adapter, "strands")
                changed = sum(r["id"].startswith("read-") and HANDLE.search(r["text"]) is not None for r in native_results(server.calls[index]["body"], protocol))
                test.assertEqual(report.replacement_count, changed, "report counts the native provider's actual view")
                test.assertIn(report.status, ("applied", "reused") if changed else ("skipped", "disabled"))
                if mode == "off":
                    test.assertEqual(report.status, "disabled")
        if mode == "compress" and not unwrapped:
            test.assertNotEqual(wire_source, SOURCES["read_logs"]); test.assertNotIn(FACT, wire_source)
            test.assertEqual(recoveries[0]["handle"], HANDLE.search(wire_source)[0])
            if method in ("resumed_session", "every_model_continuation", "model.stream"):
                test.assertTrue(any(action["type"] == "recover_resume" for action in recoveries))
            expected = 4 if parallel or method == "model.structured_output" else 5 if method in ("resumed_session", "every_model_continuation", "model.stream") else 3
        else:
            test.assertEqual(wire_source, SOURCES["read_logs"]); test.assertEqual(recoveries, [])
            expected = 3 if method in ("resumed_session", "every_model_continuation", "model.stream", "model.structured_output") else 2
            if mode == "off":
                test.assertEqual(runtime.plans, [])
        test.assertEqual(len(server.calls), expected)
        test.assertEqual(len(hooks), phase_start["calls"] if method in ("model.stream", "model.structured_output") else len(server.calls))
        test.assertEqual(server.errors, [])
        for message in agent.messages:
            key = (id(agent), id(message))
            if key in message_refs:
                test.assertEqual(message.get("tracking_id"), message_refs[key])
        if method == "every_model_continuation":
            test.assertEqual(hook_resumes, 1)
        return {"mode": "unwrapped" if unwrapped else mode, "method": method, "native_entry_point": "Model.stream + native process_stream + DecoratedFunctionTool.stream" if method == "model.stream" else method,
            "native_tool_calls": executions, "provider_calls": len(server.calls), "native_model_hook_calls": len(hooks),
            "source_sha256": digest(SOURCES["read_logs"]), "provider_source_sha256": digest(normalized(wire_source)), "transformed_request_sha256": digest(normalized(wire)),
            "omitted_fact_absent": FACT not in wire_source, "recovery_requests": len(recoveries), "recovered_source_sha256": digest(SOURCES["read_logs"]) if recoveries else None,
            "original_result_identity_preserved": True, "native_history_sha256": digest(normalized(agent.messages)), "callback_reports": [report_view(report) for report in reports],
            "native_tracking_ids_preserved": True,
            "evidence_normalization": ["opaque recovery handles", "native message tracking UUID relationships", "observed nonnegative first-byte timing"],
            "native_hook_resumes": hook_resumes, **details}
    except BaseException:
        print("CAVEMAN_MIDDLEWARE_FAILURE " + json.dumps({"family": "F09", "cell_id": cell["id"], "mode": mode,
            "lifecycle_action": lifecycle_action, "unwrapped": unwrapped, "provider_calls": len(server.calls),
            "native_tool_calls": executions, "callback_reports": [report_view(report) for report in reports], "diagnostics": diagnostics,
            "runtime_plans": [{"status": plan.status, "reason": plan.reason, "replacements": len(plan.replacements)} for _, plan in runtime.plans],
            "required_journey_complete": False}), flush=True)
        raise
    finally:
        await server.close()
        runtime.close()


async def certify_all(test):
    selected = os.environ.get("CAVEMAN_STRANDS_CERT_CELL", "")
    for cell in CELLS:
        if cell["language"] != "python" or selected and selected not in cell["id"]:
            continue
        with test.subTest(cell=cell["id"]):
            controls = {mode: await run(test, cell, mode) for mode in ("compress", "off", "outage")}
            active = controls["compress"]
            test.assertEqual(active["native_tool_calls"], controls["off"]["native_tool_calls"])
            test.assertEqual(active["native_tool_calls"], controls["outage"]["native_tool_calls"])
            if cell["method"] == "cancel_and_close":
                baseline = await run(test, cell, "off", unwrapped=True)
                for result in controls.values():
                    test.assertEqual(result["lifecycle"], baseline["lifecycle"])
                active["unwrapped_native_cancellation"] = baseline["lifecycle"]
                active["strict_transport_cancellation_passed"] = baseline["lifecycle"]["provider_closed_before_release"]
                early = {mode: await run(test, cell, mode, lifecycle_action="return") for mode in ("compress", "off", "outage")}
                early_baseline = await run(test, cell, "off", lifecycle_action="return", unwrapped=True)
                for result in early.values():
                    test.assertEqual(result["lifecycle"], early_baseline["lifecycle"])
                active["early_generator_close"] = {"unwrapped": early_baseline["lifecycle"], **early}
                active["strict_transport_early_close_passed"] = early_baseline["lifecycle"]["provider_closed_before_release"]
            observed = lambda value: {"outcome": "observed", **value}
            free = {"outcome": "recovery_free", "reason": "native Model.structured_output contract", "replacements": 0, "recovery_requests": 0,
                    "typed": active.get("typed")} if cell["structured_output"] else None
            observations = {
                "native_application": observed({"provider": cell["provider"], "method": cell["method"], "native_entry_point": active["native_entry_point"]}),
                "real_tool_result": observed({"native_tool_calls": active["native_tool_calls"], "source_sha256": active["source_sha256"], "native_result_identity_preserved": True}),
                "transformed_provider_request": free or observed({"request_sha256": active["transformed_request_sha256"], "provider_source_sha256": active["provider_source_sha256"], "omitted_fact_absent": active["omitted_fact_absent"]}),
                "omitted_fact_requested": free or observed({"native_recovery_calls": active["recovery_requests"], "exact_wire_handle_requested": True}),
                "host_executes_exact_recovery": free or observed({"source_sha256": active["source_sha256"], "exact_recovered_sha256": active["recovered_source_sha256"], "native_executor": "Strands decorated tools"}),
                "native_result_history_events_and_call_count": observed(active), "off_baseline": observed(controls["off"]), "optimizer_unavailable": observed(controls["outage"]),
            }
            for assertion, observation in observations.items():
                print("CAVEMAN_MIDDLEWARE_OBSERVATION " + json.dumps({"cell_id": cell["id"], "test_id": TEST_ID, "assertion": assertion, "observation": observation}), flush=True)
